src/canopy/llm/openai.py

```diff
@@ -164,17 +159,24 @@ def enforced_function_call(self,
 
         messages = [m.dict() for m in messages]
 
-        chat_completion = openai.ChatCompletion.create(
+        # this enforces the model to call the function
+        function_call = completion_create_params.ChatCompletionFunctionCallOptionParam(
+            name=function.name)
+
+        chat_completion = self._client.chat.completions.create(
             model=self.model_name,
             messages=messages,
-            functions=[function.dict()],
+            functions=[completion_create_params.Function(
+                name=function.name,
+                description=function.description,
+                parameters=function.parameters.dict())],
             function_call=function_call,
             max_tokens=max_tokens,
             **model_params_dict
         )
```

Comment:
From the docs - I think you also need to pass `response_format={ "type": "json_object" }`.  
Or is that guaranteed \ automatic if `function_call` ( or `tool_choice`) is True?